### PR TITLE
Add parametrization for normalization

### PR DESCRIPTION
**Pull request by y-prudent**
**Prerequisite:** [Torch parametrization tutorial](https://pytorch.org/tutorials/intermediate/parametrizations.html )

## Features

- Rewrite bjorck, frobenius, and lconv normalizations so that they use torch parametrization instead of forward pre hooks. 
- Faster inference in the `model.eval()` mode (when `model.training` is set to `False`, spectral and bjorck normalizations use cached tensors to perform free normalizations).
- Vanilla model conversion can be done on any model (not only Sequential) with `vanilla_model`. Be careful, this is an in-place conversion!
- Torch `parametrize.cached()` feature is now also usable on Lipschitz layers, allowing to save memory and compute when the same kernel is applied multiple times in an inference step (very useful for RNNs, multi-level convolutions, etc.). Here is how to use it:

```python
import torch.nn.utils.parametrize as P

with P.cached():
    y1 = lip_layer(x1)    # at the first call, compute normalization and save the reparametrized weights
    y2 = lip_layer(x2)    # at the second call, reuse saved reparametrized weights
```

### ⚠️ Important note:

Models using parametrized modules can only be serialized through `state_dict()`. So `torch.save(model, PATH)` is not possible anymore and will raise an error. Instead, save and load your models like this:
```python
# save model
torch.save(
    {
        'model_state_dict': model.state_dict(),
        'model_kwargs': config,  # arguments used to build the model
    }, PATH
)

# load model
checkpoint = torch.load(PATH)
model = TheModelClass(**checkpoint["model_kwargs"])
model.load_state_dict(checkpoint['model_state_dict'])
```

For more information, [check this torch tutorial](https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-a-general-checkpoint-for-inference-and-or-resuming-training).

### TODO

- [x] Rewrite bjorck, frobenius and lconv normalizations with parametrization
- [x] Manage the suppression of individual parametrizations (torch native functions only allow the deletion of all the parametrizations at once)
- [x] Manage the parametrization of lconv_norm that depends on the input shape (=> solution: forward pre hooks + parametrization) 
- [x] Also parametrize the global lipschitz coef multiplication
- [x] Write associated tests
- [x] Remove deprecated hook relative files
- [x] Check that vanilla_export works well
- [x] Check `parametrize.cached()` feature

### CODE REVIEW

File: .github/workflows/python-lints.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       max-parallel: 4
       matrix:
-        python-version: [3.6, 3.7, 3.8]
+        python-version: [3.7, 3.8, 3.9]
```
Comment (cofri) [inline .github/workflows/python-lints.yml:11]: From the [statuses of Python versions](https://devguide.python.org/versions/), Python 3.7 is already deprecated. Maybe we can use a more powerful test matrix with more recent Python versions (and also PyTorch versions?).
It's a more important change and it could be postponed to a future PR

File: deel/torchlip/modules/module.py [hunk 2]
```diff
@@ -44,6 +44,25 @@
 logger = logging.getLogger("deel.torchlip")
 
 
+def vanilla_model(model: nn.Module):
```
Comment (cofri) [inline deel/torchlip/modules/module.py:47]: This function might be added to the `__all__` variable in `deel/torchlip/__init__.py`. It could then be accessed directly from `deel.torchlip` import.

**Review by franckma31 — APPROVED** [review]
Ok for merging

File: .github/workflows/python-lints.yml (hunk 1 above)
Comment (franckma31) [reply .github/workflows/python-lints.yml:11]: will be done in the next PR
